flight/PiOS/STM32F4xx/pios_usart.c

```diff
@@ -325,30 +370,163 @@ static void PIOS_USART_generic_irq_handler(uintptr_t usart_id)
 		}
 	}
+
+	if (sr & USART_SR_IDLE) {
+		/* If line goes idle, trigger DMA IRQ for data processing. */
+		DMA_Cmd(usart_dev->cfg->dma_recv->stream, DISABLE);
+		while (DMA_GetCmdStatus(usart_dev->cfg->dma_recv->stream) == ENABLE) ;
+	}
 	
 	/* Check if TXE flag is set */
 	bool tx_need_yield = false;
 	if (sr & USART_SR_TXE) {
 		if (usart_dev->tx_out_cb) {
-			uint8_t b;
-			uint16_t bytes_to_send;
-			
-			bytes_to_send = (usart_dev->tx_out_cb)(usart_dev->tx_out_context, &b, 1, NULL, &tx_need_yield);
-			
-			if (bytes_to_send > 0) {
-				/* Send the byte we've been given */
-				usart_dev->cfg->regs->DR = b;
+
+			if (usart_dev->dma_send_buffer) {
+				/* Here's how it goes:
+
+				   TXE interrupt triggers, because USART has nothing to do. If there's more than 1 byte to send,
+				   the TXE IRQ gets disabled and a single DMA transfer gets triggered. When that transfer is done,
+				   it'll reenable TXE IRQ, to repeat this whole thing. */
+
+				if (DMA_GetCmdStatus(usart_dev->cfg->dma_send->stream) == ENABLE) {
+					/* If we get here while DMA is enabled, for whatever reason, GTFO.
+					   Also, re-disable TXE. */
+					USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE, DISABLE);
+					return;
+				}
+
+				uint16_t bytes_to_send = (usart_dev->tx_out_cb)(usart_dev->tx_out_context, (uint8_t*)usart_dev->dma_send_buffer,
+						PIOS_USART_DMA_SEND_BUFFER_SZ, NULL, &tx_need_yield);
+
+				if (bytes_to_send > 1) {
+					USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE, DISABLE);
+
+					DMA_MemoryTargetConfig(usart_dev->cfg->dma_send->stream, usart_dev->dma_send_buffer, DMA_Memory_0);
+					DMA_SetCurrDataCounter(usart_dev->cfg->dma_send->stream, bytes_to_send);
+					DMA_ITConfig(usart_dev->cfg->dma_send->stream, DMA_IT_TC, ENABLE);
+
+					USART_DMACmd(usart_dev->cfg->regs, USART_DMAReq_Tx, ENABLE);
+					DMA_Cmd(usart_dev->cfg->dma_send->stream, ENABLE);
+				} else if (bytes_to_send == 1) {
+					/* Don't use DMA for single byte. */
+					usart_dev->cfg->regs->DR = *((uint8_t*)usart_dev->dma_send_buffer);
+				} else {
+					USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE, DISABLE);
+				}
 			} else {
-				/* No bytes to send, disable TXE interrupt */
-				USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE, DISABLE);
+				/* Send byte-by-byte. */
+				uint8_t byte;
+				uint16_t bytes_to_send = (usart_dev->tx_out_cb)(usart_dev->tx_out_context, &byte, 1, NULL, &tx_need_yield);
+
+				if (bytes_to_send > 0) {
+					/* Send the byte we've been given */
+					usart_dev->cfg->regs->DR = byte;
+				} else {
+					/* No bytes to send, disable TXE interrupt */
+					USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE, DISABLE);
+				}
 			}
+		/* if (usart_dev->dma_send_buffer) */
 		} else {
 			/* No bytes to send, disable TXE interrupt */
 			USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE, DISABLE);
 		}
 	}
 }
 
+static uintptr_t PIOS_USART_GetDevice(const struct pios_usart_cfg *cfg)
```

Comment:
I like the entire PR except this :/

Reply:
I need access to some data in the dev handle. The alternatives to ditching this one I came up with would be:
- Removing the static qualifier of the PIOS_USART_x_id fields and then extern'ing it in the handler over in board_hw_defs, to pass it as parameter.
- Implement both handlers per USART over in pios_usart and call them from board_hw_defs, similar to the regular IRQ handlers, but unassigned.
- Move stuff into the config struct (icky).

Reply:
Gonna do per-USART handlers. Hides everything from board_hw_defs.